extensions/replication/queueProcessor/QueueProcessor.js

```diff
@@ -219,50 +221,58 @@ class QueueProcessor {
-            VersionId: entry.getEncodedVersionId() });
-        const incomingMsg = req.createReadStream();
-        req.on('error', err => {
-            this.log.error('error response getting data from S3',
-                           { entry: entry.getLogInfo(),
-                             origin: this.sourceConfig.s3,
-                             error: err.message,
-                             httpStatus: err.statusCode });
-            incomingMsg.emit('error', err);
-        });
-        return cb(null, incomingMsg);
-    }
-
-    _putData(entry, sourceStream, cb) {
-        this.log.debug('putting data', { entry: entry.getLogInfo() });
-        const cbOnce = jsutil.once(cb);
-        sourceStream.on('error', err => {
-            this.log.error('error from source S3 server',
-                           { entry: entry.getLogInfo(),
-                             error: err.message });
-            return cbOnce(err);
-        });
-        this.backbeatDest.putData({
-            Bucket: entry.getBucket(),
-            Key: entry.getObjectKey(),
-            CanonicalID: entry.getOwnerCanonicalId(),
-            ContentLength: entry.getContentLength(),
-            ContentMD5: entry.getContentMD5(),
-            Body: sourceStream,
-        }, (err, data) => {
+    _getAndPutData(sourceEntry, destEntry, cb) {
+        this.log.debug('getting data', { entry: sourceEntry.getLogInfo() });
+        const locations = sourceEntry.getLocation();
+        return async.mapLimit(locations, LIMIT, (part, done) => {
+            if (sourceEntry.getDataStoreETag(part) === undefined) {
+                this.log.error('cannot replicate object without ' +
+                               'dataStoreETag property',
+                               { entry: destEntry.getLogInfo() });
+                return done(errors.InvalidObjectState);
+            }
+            const req = this.S3source.getObject({
+                Bucket: sourceEntry.getBucket(),
+                Key: sourceEntry.getObjectKey(),
+                VersionId: sourceEntry.getEncodedVersionId(),
+                PartNumber: sourceEntry.getPartNumber(part),
+            });
+            const incomingMsg = req.createReadStream();
+            req.on('error', err => {
+                this.log.error('error response getting data from S3',
+                               { entry: sourceEntry.getLogInfo(),
+                                 origin: this.sourceConfig.s3,
+                                 error: err, httpStatus: err.statusCode });
+                incomingMsg.emit('error', err);
+            });
+            this.log.debug('putting data', { entry: destEntry.getLogInfo() });
+            incomingMsg.on('error', err => {
+                this.log.error('error from source S3 server',
+                               { entry: destEntry.getLogInfo(), error: err });
+                return done(err);
+            });
+            return this.backbeatDest.putData({
```

Comment:
I don't think that works, because as far as I can remember the `response` event is only emitted when all the data has been received from the source, which means we'll get stuck waiting for it if we don't start writing to the destination.

Ideally we would hook when only the headers have been received from S3 but I don't know if this is feasible.